Mozillians_Repositioning.md

```diff
@@ -0,0 +1,233 @@
+- new-front end renders search results and updates the search bar to reflect the proper search syntax
+
+##### …
+
+
+### Search Service
+
+The search service will be built on top of elastic search. It should provide the following APIs:
+
+- query for an individual profile by username / user id and return the according profile data
+- a search that exposes most of ES syntax that returns results in one of two formats:
+  - list of usernames / user ids
+  - list of minimal profile data (everything needed to render a card)
+- suggestion / auto-complete for search
+
+The search service will only be accessible by Mozillians, profile publisher and integrity checker.
+Ideally we store raw profile v2 data and just add some prefixed properties if necessary.
+
+#### Profile Updates
+
+The search service will have an endpoint to receive profile updates. This endpoint will insert or
+update the profile it received.
+
+#### Privacy Levels
+
+The search service will be responsible of filtering fields according to privacy settings. We have
+two proposals on how we can achieve this.
+
+1. We use one index per privacy level. This enables simple search but we need to do the filtering on
+   every insertion of data.
+2. We use a single index and include privacy filters in the search query and do filter fields in the
+   result set. This enables simple insertion but requires more work during search time.
+
+
+### Profile Publisher
+
+The profile publisher is the central broker for dealing with profile updates. Its main
+responsibilities are:
+
+- receive user created updates from Mozillians and send them to CIS
+  - the request/respond cycle will wait until the profile was successfully updated in CIS or an
+    error/timeout occurred
+- receive update events from CIS and send profile updates to [search service](#search-service)
+  and [orgchart service](#orgchart-service)
+
+The profile publisher will be implemented in an async mindset. Since it's mainly IO bound this
+will allow us to scale vertically for a long time.
+
+
+### Orgchart Service
+
+The orgchart service will serve any information necessary to render (parts of) the organizational
+structure of Mozilla.
+
+It's to be determined what technology will back this. We will focus on having a stable API.
+
+It exposes the following data profile v2 properties:
+- `user_id`
+- `first_name`
+- `last_name`
+- `business_title`
+- `team`
+- `company`
+- `office_location`
+
+
+## Integrity Checker
+
+In order to verify the integrity of the profile data in [search service](#search-service) and
+[orgchart service](#orgchart-service) we need a service that does either:
+
+- continuously checks random profiles
+- periodically checks all profiles
+
+This can be done by having a separate endpoint that provides a signature / checksum of profiles.
+
+If any integrity errors occur at any time this should raise an alarm. In order to resolve the issue
+we re-populate all data in the downstream services.
+
```

Comment:
can we check on-access?
ie:
attempt to access profile X:
Scenario 1)
- no cache available
- fetch profile from person-api
- signature check (which is a signature of a checksum of the data)
- use profile data

Scenario 2)
- hit cached profile
- signature check 
- valid => use profile

Scenario 3)
- hit cached or uncached profile
- signature validation fails
- re-fetch
- succeed or fail hard at this point

vs periodic crawling, which means incorrect data may still be presented or used until its crawled